### DISCUSSION TITLE
[BUG] Crashes on iOS

### DISCUSSION THREAD

**Discussion by scambier** [scambier/obsidian-omnisearch#152] [General]
**Problem description:**

Omnisearch causes an Obsidian crash on iOS

![image](https://user-images.githubusercontent.com/3216752/198985290-08d7aceb-5e45-4407-9a98-f3ee1864cb52.png)

![image](https://user-images.githubusercontent.com/3216752/198985104-f010947f-66f4-46b9-923e-84fdb8be30e7.png)

**Your environment:**

- Omnisearch version: 1.7.8
- Obsidian version:
- Operating system: iOS
- Number of notes in your vault (approx.):
- Other plugins that may be related to the issue:

**Comment by scambier** [comment]
The PDF indexing is disabled on iOS until a fix is published.

If anyone with an iOS device wants to volunteer for testing and reporting, I can make some customs builds. If someone wants to work on a PR, I'll happily guide them, too.

**Comment by bramses** [comment]
A few initial thoughts:

AFAIK Web Workers aren't fully supported on iOS devices  ([see example here](https://developer.apple.com/forums/thread/650317), or [here](https://www.quora.com/Can-you-use-WebAssembly-to-create-native-iOS-and-Android-apps) or [here](https://shareup.app/blog/using-webassembly-on-ios/)) natively which might cause errors in https://github.com/scambier/obsidian-omnisearch/blob/342cdabdd9f27cd3f7877a36c0a076a78c9104b5/src/lib.rs#L8

PDF.js has had trouble recently on iOS: https://github.com/mozilla/pdf.js/issues/15543

I couldn't find where the normal indexing step is, but perhaps the PDF indexer is competing for resources?

**Comment by scambier** [comment]
@bramses thanks for your references 👍

I guess the crashes are caused by a too high memory usage. As simple as that. There's not a single root cause. Omnisearch uses **a lot** of memory during indexing/refreshing. It eventually goes back down, ~~but all documents still need to live in RAM to update the search index when users modify their notes~~ 2022-12-03: this is no longer true, so it frees some memory and cache space.

Using my test vault as a reference: 1300 notes + 4000 PDFs (several novels)

- Startup RAM: ~250MB
- Peak memory uses: ~700MB when indexing PDFs, ~1300MB when updating the IndexedDB.
- Settled: ~550MB

Without the PDFs:

- Startup RAM: ~250MB
- Peak memory use: ~400MB when updating the IndexedDB.
- Settled: ~270MB

Honestly at that point I'm kinda fed up with this. I don't own an iOS device, and even if I did it won't allow me to diagnose anything. It's all trial and error and I think the solution will be to disable Omnisearch entirely to avoid crashes.

**Comment by scambier** [comment]
Update: the culprit is most likely the cache (de)serialization. I disabled the cache for iOS devices, and it looks like it stopped the crashes for at least one user. While I'm not satisfied with this solution, all other workarounds I tried were not better, and would have made cache management more complex. 

If your device keeps crashing, please add your input here with as much information as possible (number of notes, pdfs, copy of the console log). Thank you.

**Comment by bramses** [comment]
@scambier great find! These problems at this scale are very complex -- and break often, I imagine it will always be a WiP as devices evolve with time. I don't have any PDFs in my vault, but if I come across this bug, I'll follow up here. Thank you for the fix!

**Comment by scambier** [comment]
Update: I removed PDF indexing for _all_ mobile devices (incl. Android), which should also reduce the risk of a crash. Even if it works technically, it's not worth the headaches.

I prefer to propose a slightly inferior, but stable version of Omnisearch mobile, than something unstable.

**Comment by scambier** [comment]
I don't think I can do much more for this issue, since the crashes are more than likely caused by memory consumption.  I'm doing my best to keep it as low as possible without compromising performances too much, but Minisearch _does_ use a lot of RAM 🤷‍♂️

**Comment by carlosbaraza** [comment]
Mobile is a pretty key part of the Obsidian experience for many users. It seems that a search plugin that "just works" should probably work on iOS/Android, so I imagine we should not discard it yet. Building a project like this is a lot of work and I genuinely appreciate it.

A few ideas to support mobile:
- Could it be possible to index more slowly without consuming so many resources?
- Could we define a path to store an export of the IndexedDB? A few libraries claim to help doing this export (e.g. [indexeddb-export-import](https://github.com/Polarisation/indexeddb-export-import)). This could be done in the desktop device, and mobile could restore the backup. Inicially OCR could be disabled in mobile, but at least some PDFs would be searchable. 
- An alternative could also be to split the index generation in two parts:
  1. Extract the OCR into files. An option could be given to store them in the same path than the original PDF.
  2. Use the extracted OCR files as the source of the index
  3. Only run OCR for files that don't have already a file generated with the OCR.
  4. In mobile, the OCR wouldn't even need to run, just the indexing into the database, like it currently does.

I believe the last one could be the most flexible, and full OCR would not need to be done on each device.

**Reply by scambier** [reply]
@carlosbaraza 

> Mobile is a pretty key part of the Obsidian experience for many users. It seems that a search plugin that "just works" should probably work on iOS/Android, so I imagine we should not discard it yet.

I agree, and believe me when I say I'm not happy with this. But I've already spent a lot of time on this, and as I don't own an iOS device myself, I can't do very specific tests and rely on benevolent users.

As far as I understand, there are 2 causes for the crashes:
- Indexing a "big" vault makes Obsidian crash after a few seconds. 
- Loading a "big" cache makes Obsidian crash instantly at startup. It looks like some vaults can be indexed correctly, and only the cache loading triggers a crash.

The 2d point makes me think it's an iOS security trigger: loading the cache does not consume more memory than indexing, but consumes it much faster. I also disabled caching on iOS, because if Obsidian crashes at startup, it's extra difficult to fix for non-technical users.

**About your suggestions - if a developer with an iOS device is willing to tackle this, I'll be happy to guide them and answer their questions.**

Text extraction is only tangentially related, though using metadata files instead of the IndexedDB is an excellent idea.

**Reply by carlosbaraza** [reply]
I own an iOS device, although I'm not sure I'll have enough time to dig into the problem. Maybe the last week of the year I get some time. Do you have a contributing guideline? How do you install the plugin under development on iOS to test it out?

About owning an iOS device. If you own a Macbook, maybe you could run the [Xcode iOS Simulator](https://developer.apple.com/library/archive/documentation/IDEs/Conceptual/iOS_Simulator_Guide/GettingStartedwithiOSSimulator/GettingStartedwithiOSSimulator.html)

**Reply by scambier** [reply]
There's a [basic contributing guide](https://github.com/scambier/obsidian-omnisearch/blob/master/CONTRIBUTING.md) but it's pretty superficial and focused on UI and the search itself.

The cache loading/indexing starts here:
https://github.com/scambier/obsidian-omnisearch/blob/ed646eba7fee7874f2e22ef7d5c6b89ba544098f/src/main.ts#L120

Indexing into minisearch engine is done here:
https://github.com/scambier/obsidian-omnisearch/blob/ed646eba7fee7874f2e22ef7d5c6b89ba544098f/src/search/omnisearch.ts#L97

**Comment by hsiktas** [comment]
Just a heads-up, but in 2025, this caused issues on my 6 GB RAM iPhone. Disabling this plugin reduced the idle RAM usage on the desktop app from ~1.1 GB to ~600 MB, and the iOS app doesn't crash and restart all the time anymore.

**Reply by hsiktas** [reply]
I removed the plugin a few weeks ago, so I reinstalled it again to give the latest numbers:

- before installing this plugin: 
  - idling at ~900 MB private bytes
- after installing then plugin: 
  - indexing triggers, and it balloons up to 3.1 GB
  - then it idles again at ~1 GB
 
<img width="863" height="1025" alt="image" src="https://github.com/user-attachments/assets/cc9710bf-4e44-40b5-85ca-2caba5988eed" />

I think this is enough for iOS to even trigger the termination of a foreground process.
